src/config.c

```diff
@@ -7,105 +7,157 @@
+						omp_thread_count);
+		return REDISMODULE_ERR;
 	}
 
-	// Sanity.
-	assert(threshold > 0 && "A positive integer is required for the number of entities in virtual key");
-	RedisModule_Log(ctx, "notice", "Entities limit per graph meta key set to %d.", threshold);
+	// Set the OpenMP thread count in the configuration.
+	config.omp_thread_count = omp_thread_count;
+
+	return REDISMODULE_OK;
+}
+
+// If the user has specified a maximum number of entities per virtual key, update the configuration.
+// Returns REDISMODULE_OK on success and REDISMODULE_ERR if the argument was invalid.
+static int _Config_SetVirtualKeyEntitiesThreshold(RedisModuleCtx *ctx,
+												  RedisModuleString *entity_count_str) {
+	long long entity_count;
+	int res = _Config_ParsePositiveInteger(entity_count_str, &entity_count);
+	// Exit with error if integer parsing fails.
+	if(res != REDISMODULE_OK) {
+		const char *invalid_arg = RedisModule_StringPtrLen(entity_count_str, NULL);
+		RedisModule_Log(ctx, "warning", "Could not parse virtual key size argument '%s' as an integer",
+						invalid_arg);
+		return REDISMODULE_ERR;
+	}
+
+	// Log the new entity threshold.
+	RedisModule_Log(ctx, "notice", "Max number of entities per graph meta key set to %lld.",
+					entity_count);
+
+	// Update the entity count in the configuration.
+	config.vkey_entity_count = entity_count;
 
-	return (uint64_t)threshold;
+	return REDISMODULE_OK;
 }
 
-void Config_Init(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
-	config.vkey_entity_count = _Config_GetVirtualKeyEntitiesThreshold(ctx, argv, argc);
-	config.thread_count = _Config_GetThreadCount(ctx, argv, argc);
-	_initialized = true;
+// Initialize every module-level configuration to its default value.
+static void _Config_SetToDefaults(void) {
+	// The thread pool's default size is equal to the system's number of cores.
+	int CPUCount = sysconf(_SC_NPROCESSORS_ONLN);
+	config.thread_count = (CPUCount != -1) ? CPUCount : 1;
+
+	// Allow GraphBLAS to set the OpenMP thread count by default.
+	config.omp_thread_count = -1;
+
+	if(!Redis_Version_GreaterOrEqual(6, 0, 0)) {
+		// For redis-server versions below 6.0.0, we will not split the graph into virtual keys.
+		config.vkey_entity_count = VKEY_ENTITY_COUNT_UNLIMITED;
+	} else {
+		// The default entity count of virtual keys for server versions >= 6 is set by macro.
+		config.vkey_entity_count = VKEY_MAX_ENTITY_COUNT_DEFAULT;
+	}
+}
+
+// If an argument is a key-value pair, emit an error if we try to read past the end of the argv array.
+#define VALIDATE_KEYVAL_ARGC { \
+	if (cur >= argc) { \
+		RedisModule_Log(ctx, "warning", "Module parameter '%s' requires a value", param); \
+		return REDISMODULE_ERR; \
+	} \
+}
+
+int Config_Init(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
+	// Initialize the configuration to its default values.
+	_Config_SetToDefaults();
+
+	int res = REDISMODULE_OK;
+	for(int cur = 0; cur < argc; cur ++) {
```

Comment:
loop from `0` to `argc -1`,
that way you guarantee access to `cur + 1`
this will also make VALIDATE_KEYVAL_ARGC obsolete.

Reply:
My thinking with this setup was that we might want to have flags among our configurations instead of just key-value pairs, e.g. `disable-transposed-matrices` or `safemode`. We can restrict ourselves to always expecting key-value pairs, though.